Step-through debugging via a UART on ARM Cortex-M MCUs (#120)
memfault/interrupt: _drafts/debug-monitor.md

  volatile uint32_t *shpr3 = (uint32_t *)0xE000ED20;
  *shpr3 = 0xff;

  EXAMPLE_LOG("Monitor Mode Debug Enabled!");
  return true;
}
```

## A minimal DebugMonitor Handler

We can use the same handler we put together in [our post about fault handling]({% post_url
2019-11-20-cortex-m-fault-debug %}#halting--determining-core-register-state) to dump register state
about the code which caused the DebugMonitor exception to be invoked:

```c
typedef struct __attribute__((packed)) ContextStateFrame {
  uint32_t r0;
  uint32_t r1;
  uint32_t r2;
  uint32_t r3;
  uint32_t r12;
  uint32_t lr;
  uint32_t return_address;
  uint32_t xpsr;
} sContextStateFrame;

void debug_monitor_handler_c(sContextStateFrame *frame) {
  // DebugMonitor Exception Logic
  // ...
}

__attribute__((naked))
void DebugMon_Handler(void) {
  __asm volatile(
      "tst lr, #4 \n"
      "ite eq \n"
      "mrseq r0, msp \n"
      "mrsne r0, psp \n"
      "b debug_monitor_handler_c \n");
}
```

### Debug Fault Status Register, DFSR, 0xE000ED30

When the DebugMonitor Exception fires, the DFSR can be inspected for information about what debug
event took place:

![]({% img_url debugmon/dfsr.png %})

> Some of the events are only possible when using an **halting debug**. For the DebugMonitor the

Reviewer: ```suggestion
Some of the events are only possible when using a **halting debug**. For the DebugMonitor the
```

$ JLinkGDBServer  -if swd -device nRF52840_xxAA  -nogui
```

Then you need to compile the example application and flash it

```bash
$ cd ${INTERRUPT_REPO}/example/debugmon/
$ make
Compiling src/debug_monitor_exception.c
[...]
Linking library
Generated build/nrf52.elf
$ arm-none-eabi-gdb-py --eval-command="target remote localhost:2331"  --ex="mon reset" --ex="load"
--ex="mon reset"  --se=build/nrf52.elf
(gdb)
```

From GDB. we will disable halting debug so the program will continue running:

```bash
(gdb) set *(uint32_t*)0xE000EDF0=(0xA05F<<16)
```

In a final terminal we will open a serial connection using miniterm:

```bash
$ miniterm.py - 115200 --raw
--- Available ports:
...
---  4: /dev/cu.usbmodem0006838593731 'J-Link'
--- Enter port index or full name: 4
--- Miniterm on /dev/cu.usbmodem0006838593731  115200,8,N,1 ---
--- Quit: Ctrl+] | Menu: Ctrl+T | Help: Ctrl+T followed by Ctrl+H ---
shell> debug_mon_en
Monitor Mode Debug Enabled!
shell> bkpt
DebugMonitor Exception
DEMCR: 0x01010000
DFSR:  0x00000002 (bkpt=1, halt=0, dwt=0)
Register Dump
 r0  =0x00000001
 r1  =0x20003310
 r2  =0x73734900
 r3  =0x000009d5
 r12 =0xffffffff
 lr  =0x0000094b
 pc  =0x000009d4
 xpsr=0x210e0000

> Breakpoint Detected, Awaiting 'c' or 's'

Reviewer: Technically this line about "Breakpoint Detected" is not implemented until later in the post.

```

As expected we see bit 1 in `DFSR` is set indicating a breakpoint debug event was generated.
We can use GDB to inspect the instruction located at`0x000009d4` and confirm it is indeed a breakpoint
instruction:

```
(gdb) x/i 0x000009d4
  0x9d4 <prv_issue_breakpoint>:	bkpt	0x0001
```

We can also dump the instruction and confirm it matches the BKPT instruction encoding (`0xbe..`):

![]({% img_url breakpoint/bkpt-instruction-cortex-m.png %})

```
(gdb) p/x *(uint16_t*)0x000009d4
$4 = 0xbe01
```

### Programmatically installing Breakpoints with the FPB

> For the NRF52, we can install a breakpoint using the following logic:

Reviewer: It would be nice to add a paragraph here on how the FP_COMP register works. No need to go into huge details, but 1-2 sentences would be good.

```bash
shell> help
bkpt: Issue a Breakpoint Instruction
debug_mon_en: Enable Monitor Debug Mode
fpb_dump: Dump Active FPB Settings
fpb_set_breakpoint: Set Breakpoint [Comp Id] [Address]
call_dummy_funcs: Invoke dummy functions
dump_dummy_funcs: Print first instruction of each dummy function
help: Lists all commands
```

Let's try configuring a breakpoint for `dummy_function_4`, address `0x270`, and see what happens:

```bash
shell> dump_dummy_funcs
 ...
dummy_function_4: Starts at 0x270. First Instruction = 0x48024901
 ...
shell> debug_mon_en
Monitor Mode Debug Enabled!

shell> fpb_set_breakpoint 0 0x270
Set breakpoint on address 0x270 in FP_COMP[0] Succeeded

shell> fpb_dump
FPB Revision: 0, Enabled: 1, Hardware Breakpoints: 6
  FP_COMP[0] Enabled 1, Replace: 1, Address 0x270
  ...

shell> call_dummy_funcs
stub function 'dummy_function_1' called
stub function 'dummy_function_2' called
stub function 'dummy_function_3' called
DebugMonitor Exception
DEMCR: 0x01010000
DFSR:  0x00000002 (bkpt=1, halt=0, dwt=0)
Register Dump
 r0  =0x200031d8
 r1  =0x00000028
 r2  =0x00000001
 r3  =0x00000271
 r12 =0xffffffff
 lr  =0x000009ed
 pc  =0x00000270
 xpsr=0x010f0000

> Debug Event Detected, Awaiting 'c' or 's'

Reviewer: same here with "Debug Event Detected".

```

As expected we see our `DebugMon_Handler` has been invoked on address `pc =0x00000270` which
is where we installed the breakpoint!

#### Adding a CLI to the DebugMonitor Exception

To add a CLI for DebugMonitor handling, all we need is a way to read in bytes over a UART from a ISR. This
can be done by either polling the peripheral or by using an interrupt that is configured to run at
a higher priority than the DebugMonitor exception to do the reading.

> 

Reviewer: Mention the shell library?

Author: We'd have to refactor the shell a little bit to support this use case. There's only one context so it would not be safe to have the user land shell and the ISR shell share the same thing

```bash
shell> debug_mon_en
Monitor Mode Debug Enabled!

shell> fpb_set_breakpoint 0 0x270

shell> call_dummy_funcs
stub function 'dummy_function_1' called
stub function 'dummy_function_2' called
stub function 'dummy_function_3' called
DebugMonitor Exception
DEMCR: 0x01010000
DFSR:  0x00000002 (bkpt=1, halt=0, dwt=0)
Register Dump
 r0  =0x200031d8
 r1  =0x00000028
 r2  =0x00000001
 r3  =0x00000271
 r12 =0xffffffff
 lr  =0x00000a05
 pc  =0x00000270
 xpsr=0x010d0000
Debug Event Detected, Awaiting 'c' or 's'
# Type 's'
Got char 's'!
 ...
 pc  =0x00000272
 ...
Debug Event Detected, Awaiting 'c' or 's'
# Type 's'
Got char 's'!
 ...
 pc  =0x00000274
 ...
Debug Event Detected, Awaiting 'c' or 's'
# Type 's'
Got char 's'!
 ...
 pc  =0x000006c0
 ...
Debug Event Detected, Awaiting 'c' or 's'
# Resume flow 'c' and we should see the
# rest of the functions get called!
Got char 'c'!

stub function 'dummy_function_4' called
stub function 'dummy_function_5' called
stub function 'dummy_function_6' called
stub function 'dummy_function_7' called
stub function 'dummy_function_8' called
stub function 'dummy_function_9' called
stub function 'dummy_function_ram' called
```

Nice, it works!

> 

Reviewer: I think there's one small thing missing from this article: a read or hexdump function of some sort so you can poke at memory. This makes the single stepping & breakpoint much more interesting.

Also cool would be a backtrace thinger, though that's probably too much work.

Author: I've added a note about how this is something someone could do. My current CLI from the ISR is super minimal (no echo, only operates on one char). I agree these are the logical things I would want to add next!

stub function 'dummy_function_9' called
stub function 'dummy_function_ram' called
```

Nice, it works!

## Applications

There's a couple projects online today that make use of the debug monitor functionality. Notably,

- SEGGER ships an SDK[^4] that you compile into your project to enable Monitor Mode Debug support
  with SEGGER's tools. There's quite a few articles about how to make use of it on the NRF5x
  series[^5]. On the NRF52, this let's you set breakpoints in application code without triggering
  asserts in the Soft Device (because all that code runs from a higher priority interrupt).
- The Monitor For Remote Inspection Project on GitHub [^6] exposes a DebugMonitor port which
  implements the GDB Remote Serial Protocol[^7]. When compiled into a project, you can connect GDB
  directly to your board via the UART connection!

There's a lot of additional use cases you may want to include an
implementation for in your system as well. Some of my favorites include:

- exposing a debug interface for third party developers
  on an embedded platform.
- debug of sealed units where there is only a UART or network interface exposed
- enabling [post-mortem analysis]({% post_url

>   2019-11-20-cortex-m-fault-debug %}#postmortem-analysis) for any code path by being able to

Reviewer: Let's remove this.

---
title: "Step-through debugging via a UART on ARM Cortex-M MCUs"
description: "An exploration of how to debug running devices by taking advantage of debug monitor
mode on ARM Cortex-M MCUs"
image: /img/debugmon/debug-via-shell.png
tag: [cortex-m]
author: chris
---

> Debugging an MCU, requires attaching a debugger via JTAG, and halting the core. Or does it?

Reviewer 2: ```suggestion
Debugging an MCU requires attaching a debugger via JTAG and halting the core. Or does it?
```

There are many situations where this debug strategy is not feasible such as on

- production units where JTAG has been disabled.
- systems where halting the core causes the subsystem you are trying to debug to

>   fault. (i.e Asserts from the NRF52 SoftDevice after halting at a breakpoint [^6])

Reviewer 2: ```suggestion
  fault. (i.e Asserts from the NRF52 SoftDevice after halting at a breakpoint[^6])
```

- remote systems where only a UART or network interface is exposed.

Fortunately, ARM Cortex-M's have us covered!

<!-- excerpt start -->

This week we explore how to debug running systems with ARM Cortex-M's DebugMonitor
exception handler. We cover how to configure the MCU to operate in this mode and walk through an
example of installing breakpoints and single-stepping on a running device!

<!-- excerpt end -->

> In my opinion, DebugMonitor handling is an underutilized feature that can be applied creatively

> > for a variety of debug use cases. Hope by the end of this article I can convince you of this as well!

Reviewer 2: ```suggestion
> for a variety of debug use cases. I hope by the end of this article I can convince you of this as well!
```

_Like Interrupt? [Subscribe](http://eepurl.com/gpRedv) to get our latest
posts straight to your mailbox_

{:.no_toc}

## Table of Contents

<!-- prettier-ignore -->
* auto-gen TOC:
{:toc}

## Basic Terminology

ARM Cortex-M's support several "levels" of debug:

- **Halting debug** - This is the typical configuration you use with a debugger like GDB. In this mode, the core is
  halted while debugging. This mode requires access to the **Debug Port** via JTAG or SWD. We've
  walked through an overview of how ARM debug interfaces work in [this article]({% post_url
  2019-08-06-a-deep-dive-into-arm-cortex-m-debug-interfaces %}).
- Tracing via the **[ITM]({% post_url 2020-06-02-profiling-firmware-on-cortex-m %})**,
  **DWT** & **ETM** - These peripherals let you stream instruction execution and data

>   accesses while the system is running. Tools such as Lauterbach's Trace32[^8] SEGGER's JTrace[^9]

Reviewer 2: ```suggestion
  accesses while the system is running. Tools such as Lauterbach's Trace32[^8] and SEGGER's JTrace[^9]
```

![]({% img_url breakpoint/bkpt-instruction-cortex-m.png %})

```
(gdb) p/x *(uint16_t*)0x000009d4
$4 = 0xbe01
```

### Programmatically installing Breakpoints with the FPB

In the FPB we can program one of the`FP_COMP` register to hold the address we would like to break
on.

The layout of the `FP_COMP` register is a little odd but basically for a given `instr_addr`, `FP_COMP[31:30]` map
`instr_addr[1:0]`, `FP_COMP[28:2]` map to `instr_addr[28:2]`, and the lowest bit `FP_COMP[0]` controls
whether the breakpoint is enabled or not.

For the NRF52, we can transcribe the logic into code as follows:

```c
bool fpb_set_breakpoint(size_t comp_id, uint32_t instr_addr) {
  if (instr_addr >= 0x20000000) {
    // for revision 1 only breakpoints in code can be installed :/
    return false;
  }
  // make sure the FPB is enabled
  FPB->FP_CTRL |= 0x3;

  const uint32_t replace = (instr_addr & 0x2) == 0 ? 1 : 2;
  const uint32_t fp_comp = (instr_addr & ~0x3) | 0x1 | (replace << 30);
  FPB->FP_COMP[comp_id] = fp_comp;
  return true;
}
```

> For more details on how to configure the Flash Patch And Breakpoint (**FPB**) unit,

> > [check out our post on the topic]({% post_url 2020-06-17-cortex-m-breakpoints %}#flash-patch--breakpoint-unit).

Reviewer 2: I'd bring the name of the post in here rather than just "check out our post on the topic"

stub function 'dummy_function_ram' called
```

Nice, it works!

> At this point, to make the DebugMonitor CLI interface more useful, we could extend the shell to
> dump the contents of the stack or other memory regions.

## Applications

There's a couple projects online today that make use of the debug monitor functionality. Notably,

- SEGGER ships an SDK[^4] that you compile into your project to enable Monitor Mode Debug support
  with SEGGER's tools. There's quite a few articles about how to make use of it on the NRF5x
  series[^5]. On the NRF52, this let's you set breakpoints in application code without triggering
  asserts in the Soft Device (because all that code runs from a higher priority interrupt).
- The Monitor For Remote Inspection Project on GitHub [^6] exposes a DebugMonitor port which
  implements the GDB Remote Serial Protocol[^7]. When compiled into a project, you can connect GDB
  directly to your board via the UART connection!

There's a lot of additional use cases you may want to include an
implementation for in your system as well. Some of my favorites include:

- exposing a debug interface for third party developers
  on an embedded platform.

> - debug of sealed units where there is only a UART or network interface exposed

Reviewer 2: ```suggestion
- debugging sealed units where there is only a UART or network interface exposed
```
